vowpalwabbit/core/src/reductions/interaction_ground.cc

```diff
@@ -3,111 +3,240 @@
-  CB::cb_class label = CB_ADF::get_observed_cost_or_default_cb_adf(ec_seq);
-  ig.total_uniform_cost += label.cost / label.probability / ec_seq.size();  //=p(uniform) * IPS estimate
-  ig.total_uniform_reward += -label.cost / label.probability / ec_seq.size();
-
-  // find prediction & update for cost
-  base.predict(ec_seq);
-  ig.total_importance_weighted_cost += get_cost_estimate(label, ec_seq[0]->pred.a_s[0].action);
-  base.learn(ec_seq);
-
-  // find prediction & update for reward
-  label.cost = -label.cost;
-  base.predict(ec_seq, 1);
-  ig.total_importance_weighted_reward += get_cost_estimate(label, ec_seq[0]->pred.a_s[0].action);
+  std::swap(igl.ik_ftrl->all->loss, igl.ik_all->loss);
+  std::swap(igl.ik_ftrl->all->sd, igl.ik_all->sd);
+
+  float fake_cost = 0.f;
+  // 4. update multi line ex label
+  if (ik_pred * 2 > 1) {
+    // TODO: get Definitely Bad label from feedback example
+    bool is_neg = 1;
+    fake_cost = -p_unlabeled_prior + is_neg * (1 + p_unlabeled_prior); // TODO: update to latest version
+  }
 
-  // change cost to reward
-  negate_cost(ec_seq);
+  // 5. Train pi policy
+  ec_seq[chosen_action_idx]->l.cb.costs[0].cost = fake_cost;
+
+  VW::reductions::swap_ftrl(igl.pi_ftrl.get(), igl.ik_ftrl);
   base.learn(ec_seq, 1);
-  negate_cost(ec_seq);
+  VW::reductions::swap_ftrl(igl.pi_ftrl.get(), igl.ik_ftrl);
+  ec_seq.push_back(feedback_ex);
 }
 
-void predict(interaction_ground& ig, multi_learner& base, VW::multi_ex& ec_seq)
+void predict(interaction_ground& igl, multi_learner& base, VW::multi_ex& ec_seq)
 {
-  // figure out which is better by our current estimate.
-  if (ig.total_uniform_cost - ig.total_importance_weighted_cost >
-      ig.total_uniform_reward - ig.total_importance_weighted_reward)
-  {
-    base.predict(ec_seq);
-  }
-  else { base.predict(ec_seq, 1); }
+  VW::reductions::swap_ftrl(igl.pi_ftrl.get(), igl.ik_ftrl);
+  base.predict(ec_seq, 1);
+  VW::reductions::swap_ftrl(igl.pi_ftrl.get(), igl.ik_ftrl);
+}
+} // namespace
+
+void copy_ftrl(ftrl* source, ftrl* destination) {
+  destination->all = source->all;
+  destination->ftrl_alpha = source->ftrl_alpha;
+  destination->ftrl_beta = source->ftrl_beta;
+  destination->data.update = source->data.update;
+  destination->data.ftrl_alpha = source->data.ftrl_alpha;
+  destination->data.ftrl_beta = source->data.ftrl_beta;
+  destination->data.l1_lambda = source->data.l1_lambda;
+  destination->data.l2_lambda = source->data.l2_lambda;
+  destination->data.predict = source->data.predict;
+  destination->data.normalized_squared_norm_x = source->data.normalized_squared_norm_x;
+  destination->data.average_squared_norm_x = source->data.average_squared_norm_x;
+  destination->no_win_counter = source->no_win_counter;
+  destination->early_stop_thres = source->early_stop_thres;
+  destination->ftrl_size = source->ftrl_size;
+  destination->total_weight = source->total_weight;
+  destination->normalized_sum_norm_x = source->normalized_sum_norm_x;
```

Comment:
Just use the copy constructor

```suggestion
  *destination = *source;
```